### PR TITLE
streams defined in service discovery metadata for each service

### PR DESCRIPTION
**Pull request by matgabriel**

### CODE REVIEW

File: stream_provider.go [hunk 1]
```diff
@@ -254,48 +259,77 @@ func ParseStreamName(fullStreamName string) (string, string) {
...
+type StreamDefinitions []*StreamDefinition
+
+func (d StreamDefinitions) StreamNames() string {
+	return d.join(func(sd *StreamDefinition) string {
+		return sd.streamName
+	})
+}
+
+func (d StreamDefinitions) join(f func(*StreamDefinition) string) string {
+	l := make([]string, len(d))
```
Comment (pdaures) [inline stream_provider.go]: nice!!!

File: stream_provider.go [hunk 2]
```diff
@@ -254,48 +259,77 @@ func ParseStreamName(fullStreamName string) (string, string) {
...
+
+func (r *streamRegistry) streams() StreamDefinitions {
+	r.Lock()
+	defer r.Unlock()
+	return extractStreamsDefinitions(r)
+}
+
+//To be called only with a lock on the stream registry
+func extractStreamsDefinitions(r *streamRegistry) StreamDefinitions {
+	result := make([]*StreamDefinition, len(r.providers))
```
Comment (pdaures) [inline stream_provider.go]: r.Lock()?
Or no multithreading issue possible here?

Comment (pdaures) [reply stream_provider.go]: EDIT: mistake, all callers call r.Lock before

File: stream_provider.go [hunk 3]
```diff
@@ -254,48 +259,77 @@ func ParseStreamName(fullStreamName string) (string, string) {
...
+	i := 0
+	for _, v := range r.providers {
```
Comment (pdaures) [inline stream_provider.go]: you can directly use:

`for i,v := range r.providers{
 }`
so you don't need to increment i by hand

**Comment by pdaures** [comment]
LGTM

File: stream_provider.go (hunk 3 above)
Comment (matgabriel) [reply stream_provider.go]: r.providers is a map, to in this case i would be the key, wouldn't it?

File: stream_provider.go (hunk 2 above)
Comment (matgabriel) [reply stream_provider.go]: this method is designed to be called from other methods that already hold a lock on r
and since Go locks are not reentrant...

**Pull request merged**